src/CrawlQueue/ArrayCrawlQueue.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Spatie\Crawler\CrawlQueue;
+
+use Psr\Http\Message\UriInterface;
+use Spatie\Crawler\CrawlUrl;
+use Spatie\Crawler\Exception\UrlNotFoundByIndex;
+use TypeError;
+
+/**
+ * Crawl queue implemented with arrays.
+ */
+class ArrayCrawlQueue implements CrawlQueue
+{
+    /**
+     * All known URLs, indexed by URL string.
+     *
+     * @var CrawlUrl[]
+     */
+    protected $urls = [];
+
+    /**
+     * Pending URLs, indexed by URL string.
+     *
+     * @var CrawlUrl[]
+     */
+    protected $pendingUrls = [];
+
```

Comment:
Add
```protected $processedUrls = []```

Reply:
There is no real justification for an additional `$processedUrls` here: processed URLs are those that are in `$urls` and not in `$pendingUrls`, the check is easy and with a very low performance cost.

OTOH, adding an extra property to store these would **double the memory cost** of the crawl queue.

What could be done is *replacing* `$urls` with `$processedUrls`; so that pending URLs are only stored in `$pendingUrls`, and moved to `$processedUrls` once done. I'm not sure if this is a big improvement, but it can be done.

Reply:
Ok, let's leave it like it is for now. Thanks for your feedback.